tiled/client/context.py

```diff
@@ -22,25 +21,78 @@
 
 USER_AGENT = f"python-tiled/{tiled_version}"
 API_KEY_AUTH_HEADER_PATTERN = re.compile(r"^Apikey (\w+)$")
-PROMPT_FOR_REAUTHENTICATION = None
 
 
-def prompt_for_credentials(username, password):
+def _choose_identity_provider(providers, provider=None):
+    if len(providers) == 1:
+        # There is only one choice, so no need to prompt the user.
+        (spec,) = providers
+    else:
+        while True:
+            print("Authenticaiton providers:")
+            for i, spec in enumerate(providers, start=1):
+                print(f"{i} - {spec['provider']}")
+            raw_choice = input(
+                "Choose an authentication provider (or press Enter to escape): "
+            )
+            if not raw_choice:
+                print("No authentication provider chosen. Failed.")
+                break
+            try:
+                choice = int(raw_choice)
+            except TypeError:
+                print("Choice must be a number.")
+                continue
+            try:
+                spec = providers[choice - 1]
+            except IndexError:
+                print(f"Choice must be a number 1 through {len(providers)}.")
+                continue
+            break
+    return spec
+
+
+def prompt_for_credentials(http_client, providers):
     """
-    Utility function that displays a username prompt.
+    Prompt for credentials or third-party login at an interactive terminal.
     """
-    if username is not None and password is not None:
-        # If both are provided, return them as-is, without prompting.
-        # This is particularly useful for GUI clients without a TTY Console.
-        return username, password
-    elif username:
-        username_reprompt = input(f"Username [{username}]: ")
-        if len(username_reprompt.strip()) != 0:
-            username = username_reprompt
-    else:
+    spec = _choose_identity_provider(providers)
+    auth_endpoint = spec["links"]["auth_endpoint"]
+    provider = spec["provider"]
+    mode = spec["mode"]
+    if mode == "password":
+        # Prompt for username, password at terminal.
```

Comment:
Should external clients also prompt for the username every time, rather than store the name of the last user as a convenience?

In other words, since Tiled python client has made this change, is it a recommendation for all Tiled clients?

Reply:
What do you have in mind when you say "external clients"?

Reply:
I was thinking of general GUI clients (Qt, JS, etc.) and also had not seen the nice summary in "Design changes" when I wrote that question...so I was guessing at the reasons for removing `username` from the function args.

I see now that the reasons were partly about confusion in the function call and partly about UX (whoops, I just typed my plaintext password into the username field).

Reply:
I suppose I was asking: _Is there a security concern for caching the username as a convenience for the user?_

It looks the answer is that it may be a behavioral concern (did the user realize that it's not the right place to enter their password) but not generally a "leakage" concern of _Who was the last user before me?_

My takeaway is that an app _could_ store the username but that would be up to the app designer, and if so then it should be clear where the password gets entered.

Reply:
Right. Fully on the same page, just reiterating:

The change in behavior in Tiled here is: we used to cache the username ("default identity") and now we only cache tokens. So, you either have a valid session of you don't, and if you don't Tiled itself will not retain any state related to some previously-used identity or session. That would be up to applications to manage as they wish, as you said.